group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -983,27 +1085,81 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        // Cancel all the timers of the departed static member.
+        cancelTimers(group.groupId(), existingStaticMember.memberId());
+    }
+
+    private boolean staticMemberReplaced(int memberEpoch, ConsumerGroupMember existingStaticMember) {
+        return memberEpoch == 0 && existingStaticMember != null && existingStaticMember.memberEpoch() == LEAVE_GROUP_STATIC_MEMBER_EPOCH;
+    }
+
     /**
      * Handles leave request from a consumer group member.
      * @param groupId       The group id from the request.
      * @param memberId      The member id from the request.
+     * @param memberEpoch   The member epoch from the request.
      *
      * @return A Result containing the ConsumerGroupHeartbeat response and
      *         a list of records to update the state machine.
      */
     private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGroupLeave(
         String groupId,
-        String memberId
+        String instanceId,
+        String memberId,
+        int memberEpoch
     ) throws ApiException {
         ConsumerGroup group = getOrMaybeCreateConsumerGroup(groupId, false);
-        ConsumerGroupMember member = group.getOrMaybeCreateMember(memberId, false);
+        ConsumerGroupMember member = memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH ?
+                group.staticMember(instanceId) :
+                group.getOrMaybeCreateMember(memberId, false);
 
-        log.info("[GroupId " + groupId + "] Member " + memberId + " left the consumer group.");
-
-        List<Record> records = consumerGroupFenceMember(group, member);
+        List<Record> records = new ArrayList<>();
+        if (memberEpoch == LEAVE_GROUP_STATIC_MEMBER_EPOCH) {
+            throwIfStaticMemberValidationFails(groupId, instanceId, member, memberEpoch, memberId);
+            log.info("[GroupId {}] Member {} with instance id {} is a static member and will not be fenced from the group",
+                    group.groupId(), member.memberId(), member.instanceId());
+            records.addAll(consumerGroupStaticMemberGroupLeave(group, member, memberId));
+        } else {
+            log.info("[GroupId {}] Member {} left the consumer group.", groupId, memberId);
+            records.addAll(consumerGroupFenceMember(group, member));
+        }
         return new CoordinatorResult<>(records, new ConsumerGroupHeartbeatResponseData()
             .setMemberId(memberId)
-            .setMemberEpoch(LEAVE_GROUP_MEMBER_EPOCH));
+            .setMemberEpoch(memberEpoch));
+    }
+
+    /**
+     * Handles the case when a static member decides to leave the group.
+     * The member is not actually fenced from the group, and instead it's
+     * member epoch is updated to -2 to reflect that a member using the given
+     * instance id decided to leave the group and would be back within session
+     * timeout.
+     *
+     * @param group       The group.
+     * @param existingStaticMember      The member.
```

Comment:
nit: Let's align the description of the params.

Reply:
done.